File: android/app/src/main/java/com/example/app/domain/navcaddy/navigation/DeepLinkBuilder.kt
package caddypro.domain.navcaddy.navigation

import caddypro.domain.navcaddy.models.IntentType
import caddypro.domain.navcaddy.models.Module
import caddypro.domain.navcaddy.models.RoutingTarget
import javax.inject.Inject

/**
 * Builds navigation routes from RoutingTargets.
 *
 * Converts the generic RoutingTarget (from intent classification) into
 * a concrete NavCaddyDestination with properly typed parameters.
 *
 * Validates target before building to ensure all required parameters are present.
 *
 * Spec reference: navcaddy-engine.md R3, navcaddy-engine-plan.md Task 11
 */
class DeepLinkBuilder @Inject constructor() {

    /**
     * Build a navigation route string from a RoutingTarget.
     *
     * @param target The routing target from intent classification
     * @return Navigation route string for NavController
     * @throws IllegalArgumentException if target is invalid or missing required parameters
     */
    fun buildRoute(target: RoutingTarget): String {
        validateTarget(target)
        val destination = buildDestination(target)
        return destination.toRoute()
    }

    /**
     * Build a typed NavCaddyDestination from a RoutingTarget.
     *
     * @param target The routing target from intent classification
     * @return Typed destination with parameters
     * @throws IllegalArgumentException if screen name is unknown
     */
    fun buildDestination(target: RoutingTarget): NavCaddyDestination {
        return when (target.module) {
            Module.CADDY -> buildCaddyDestination(target)
            Module.COACH -> buildCoachDestination(target)
            Module.RECOVERY -> buildRecoveryDestination(target)
            Module.SETTINGS -> buildSettingsDestination(target)
        }
    }

    private fun buildCaddyDestination(target: RoutingTarget): NavCaddyDestination {
        return when (target.screen) {
            "club_adjustment" -> NavCaddyDestination.ClubAdjustment(
                clubId = target.parameters["clubId"] as? String
            )

            "shot_recommendation" -> NavCaddyDestination.ShotRecommendation(
                yardage = (target.parameters["yardage"] as? Number)?.toInt(),
                lie = target.parameters["lie"] as? String,
                wind = target.parameters["wind"] as? String
            )

            "round_start" -> NavCaddyDestination.RoundStart(
                courseName = target.parameters["courseName"] as? String
            )

            "score_entry" -> NavCaddyDestination.ScoreEntry(
                hole = (target.parameters["hole"] as? Number)?.toInt()
            )

            "round_end" -> NavCaddyDestination.RoundEnd

            "weather" -> NavCaddyDestination.WeatherCheck

            "stats" -> NavCaddyDestination.StatsLookup(
                statType = target.parameters["statType"] as? String
            )

            "course_info" -> {
                val courseId = target.parameters["courseId"] as? String
                    ?: throw IllegalArgumentException("CourseInfo requires courseId parameter")
                NavCaddyDestination.CourseInfo(courseId)
            }

            else -> throw IllegalArgumentException("Unknown CADDY screen: ${target.screen}")
        }
    }

    private fun buildCoachDestination(target: RoutingTarget): NavCaddyDestination {
        return when (target.screen) {
            "drill" -> NavCaddyDestination.DrillScreen(
                drillId = target.parameters["drillId"] as? String,
                focusArea = target.parameters["focusArea"] as? String
            )

            "practice" -> NavCaddyDestination.PracticeSession

            else -> throw IllegalArgumentException("Unknown COACH screen: ${target.screen}")
        }
    }

    private fun buildRecoveryDestination(target: RoutingTarget): NavCaddyDestination {
        return when (target.screen) {
            "overview" -> NavCaddyDestination.RecoveryOverview

            "data_entry" -> NavCaddyDestination.RecoveryDataEntry(
                dataType = target.parameters["dataType"] as? String
            )

            else -> throw IllegalArgumentException("Unknown RECOVERY screen: ${target.screen}")
        }
    }

    private fun buildSettingsDestination(target: RoutingTarget): NavCaddyDestination {
        return when (target.screen) {
            "equipment" -> NavCaddyDestination.EquipmentManagement

            "settings" -> NavCaddyDestination.SettingsScreen(
                settingKey = target.parameters["settingKey"] as? String
            )

            "help" -> NavCaddyDestination.HelpScreen

            else -> throw IllegalArgumentException("Unknown SETTINGS screen: ${target.screen}")
        }
    }

    /**
     * Validate that target has valid module and screen.
     * Throws IllegalArgumentException if invalid.
     */
    private fun validateTarget(target: RoutingTarget) {
        require(target.screen.isNotBlank()) {
            "RoutingTarget must have a non-blank screen name"
        }
    }

    companion object {
        /**
         * Quick helper to validate if a target is buildable without throwing.
         * Useful for testing/debugging.
         */
        fun canBuild(target: RoutingTarget): Boolean {
            return try {
                DeepLinkBuilder().buildDestination(target)
                true
            } catch (e: IllegalArgumentException) {
                false
            }
        }
    }
}

File: ios/App/Sources/Core/Domain/NavCaddy/Navigation/DeepLinkBuilder.swift
import Foundation

/// Converts RoutingTarget to NavCaddyDestination for navigation.
///
/// Spec R3: Maps intent -> module route + required parameters
/// Validates targets before building destinations.
struct DeepLinkBuilder {
    /// Builds a NavCaddyDestination from a RoutingTarget
    ///
    /// - Parameter target: The routing target containing module, screen, and parameters
    /// - Returns: A NavCaddyDestination if the target is valid, nil otherwise
    func buildDestination(from target: RoutingTarget) -> NavCaddyDestination? {
        // Match on module and screen to determine destination
        switch (target.module, target.screen) {
        // MARK: - Caddy Module

        case (.caddy, "club_adjustment"):
            guard let club = target.parameters["club"] else {
                return nil // Club parameter is required
            }
            return .clubAdjustment(club: club)

        case (.caddy, "shot_recommendation"):
            let yardage = target.parameters["yardage"].flatMap { Int($0) }
            let club = target.parameters["club"]
            let lie = target.parameters["lie"]
            let wind = target.parameters["wind"]
            return .shotRecommendation(yardage: yardage, club: club, lie: lie, wind: wind)

        case (.caddy, "weather_check"):
            let location = target.parameters["location"]
            return .weatherCheck(location: location)

        case (.caddy, "course_info"):
            guard let courseId = target.parameters["course_id"] else {
                return nil // Course ID is required
            }
            return .courseInfo(courseId: courseId)

        case (.caddy, "round_start"):
            let courseId = target.parameters["course_id"]
            return .roundStart(courseId: courseId)

        case (.caddy, "score_entry"):
            guard let holeStr = target.parameters["hole"],
                  let hole = Int(holeStr) else {
                return nil // Hole number is required
            }
            return .scoreEntry(hole: hole)

        case (.caddy, "round_end"):
            return .roundEnd

        // MARK: - Coach Module

        case (.coach, "drill_screen"):
            let drillType = target.parameters["drill_type"]
            let focusClub = target.parameters["club"]
            return .drillScreen(drillType: drillType, focusClub: focusClub)

        case (.coach, "stats_lookup"):
            let statType = target.parameters["stat_type"]
            let dateRange = target.parameters["date_range"]
            return .statsLookup(statType: statType, dateRange: dateRange)

        case (.coach, "pattern_query"):
            let club = target.parameters["club"]
            let pressureContext = target.parameters["pressure_context"]
            return .patternQuery(club: club, pressureContext: pressureContext)

        // MARK: - Recovery Module

        case (.recovery, "overview"):
            return .recoveryOverview

        case (.recovery, "detail"):
            guard let date = target.parameters["date"] else {
                return nil // Date is required
            }
            return .recoveryDetail(date: date)

        // MARK: - Settings Module

        case (.settings, "equipment_info"):
            let clubToEdit = target.parameters["club"]
            return .equipmentInfo(clubToEdit: clubToEdit)

        case (.settings, "settings"):
            let section = target.parameters["section"]
            return .settings(section: section)

        case (.settings, "help"):
            return .help

        // MARK: - Unknown/Unsupported

        default:
            // Log unknown screen for debugging
            print("[DeepLinkBuilder] Unknown routing target: module=\(target.module.rawValue), screen=\(target.screen)")
            return nil
        }
    }

    /// Validates a routing target before building
    ///
    /// - Parameter target: The routing target to validate
    /// - Returns: true if the target has valid structure
    func validate(_ target: RoutingTarget) -> Bool {
        // Screen name should not be empty
        guard !target.screen.isEmpty else {
            return false
        }

        // Module should be valid (already enforced by enum)
        // Check for required parameters based on screen
        switch (target.module, target.screen) {
        case (.caddy, "club_adjustment"):
            return target.parameters["club"] != nil

        case (.caddy, "course_info"):
            return target.parameters["course_id"] != nil

        case (.caddy, "score_entry"):
            if let holeStr = target.parameters["hole"],
               let hole = Int(holeStr),
               hole >= 1 && hole <= 18 {
                return true
            }
            return false

        case (.recovery, "detail"):
            return target.parameters["date"] != nil

        default:
            // Most screens have optional parameters
            return true
        }
    }
}
